### PR TITLE
Test coverage for Promises libraries

### PR DESCRIPTION
**Pull request by dbukhmastov**
Test development according to https://electricimp.atlassian.net/browse/CSE-310

### CODE REVIEW

**Comment by snork-nbl** [comment]
A general comment: please make sure you have all the EI copyrights in place.

File: tests/reject/ManyResolveRejectAllFuncs.agent.test.nut [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        local cleanPath = @(p) p.len() == 0 ? p : p.slice(1);
+        if (level > 32) {
+            server.log("Possible cyclic reference at " + cleanPath(path));
+            return false;
+        }
+        switch (type(value1)) {
+            case "table":
+            case "class":
+            case "array":
+                foreach (k, v in value1) {
```
Comment (snork-nbl) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: You have two arrays: [1,1,2] and [1,2,2]. This comparison returns true. Is it an expected behavior?

File: tests/PendingState.agent.test.nut [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ */
+
+class PendingState extends ImpTestCase {
+    /**
+     * race  - then
+     */
+    function testRaceThen() {
+        return Promise(function(ok, err) {
+            local isResolved = false;
+            ::Promise.race([
```
Comment (snork-nbl) [inline tests/PendingState.agent.test.nut]: I believe you don't actually need to use :: here. It should be accessible even without it.

File: tests/PendingState.agent.test.nut [hunk 3]
```diff
@@ -0,0 +1,171 @@
+/**
```
Comment (snork-nbl) [inline tests/PendingState.agent.test.nut]: I don't want you to push too hard here, but according to https://electricimp.com/docs/squirrel/codestyle/, you shouldn't be using multiline comments here.

File: tests/reject/ManyResolveRejectAllFuncs.agent.test.nut [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    * @param {string} message
+    * @param {boolean} isForwardPass - on forward pass value1 is treated "expected", value2 as "actual" and vice-versa on backward pass
+    * @param {string} path - current slot path
+    * @param {int} level - current depth level
+    * @private
+    */
+    function _assertDeepEqualImpl(value1, value2, message, isForwardPass, path = "", level = 0) {
+        local result = true;
+        local cleanPath = @(p) p.len() == 0 ? p : p.slice(1);
+        if (level > 32) {
```
Comment (snork-nbl) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: uhh

File: tests/reject/ManyResolveRejectAllFuncs.agent.test.nut (hunk 1 above)
Comment (dbukhmastov) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: server.log("[1,1,2] and [1,2,2] "+_assertDeepEqual([1,1,2],[1,2,2]));
Log:
2017-03-02 17:45:24+0000 [Agent]  [1,1,2] and [1,2,2] false

Comment (snork-nbl) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: My bad, yes it should work, you are right!

File: tests/PendingState.agent.test.nut (hunk 2 above)
Comment (dbukhmastov) [inline tests/PendingState.agent.test.nut]: local isResolved = false;
            Promise.race([
The result is 
[+56.74/0.00s test] PendingState::testRaceThen()
[+56.74/0.00s test] Failure: the index 'race' does not exist (line 17)

File: tests/PendingState.agent.test.nut (hunk 3 above)
Comment (dbukhmastov) [inline tests/PendingState.agent.test.nut]: Done. multiline comments have been replaced with single-line in all tests. The copyrights have been added to all tests.

File: tests/PendingState.agent.test.nut (hunk 2 above)
Comment (snork-nbl) [inline tests/PendingState.agent.test.nut]: so it finds Promise, but doesn't find race. hmm, do you understand why?

Comment (dbukhmastov) [inline tests/PendingState.agent.test.nut]: No, but There are two Promise in scope
The first is generated by impTest:
function __module_ImpUnit_Promise() {
  local exports = class {...}
  return exports;
}
__module_ImpUnit_Promise_exports <- __module_ImpUnit_Promise();
Promise <- __module_ImpUnit_Promise_exports;

The second is from promise.class.nut :
.... * @version 3.0.0....
class Promise {.....

So the "::Promise" is tested Promise, the "Promise" is service Promise for testing process.

File: tests/reject/ManyResolveRejectAllFuncs.agent.test.nut (hunk 4 above)
Comment (dbukhmastov) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: Done. The MAX depth level is set to 2

Comment (snork-nbl) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: no-no, this is not what I mean. probably 32 is fine. but in general it looks strange that one needs to invent this bicycle...

File: tests/PendingState.agent.test.nut (hunk 2 above)
Comment (snork-nbl) [inline tests/PendingState.agent.test.nut]: Got it. Whatever...

File: tests/reject/ManyResolveRejectAllFuncs.agent.test.nut (hunk 4 above)
Comment (dbukhmastov) [inline tests/reject/ManyResolveRejectAllFuncs.agent.test.nut]: duplicate code has been replaced with _assertDeepEqualWrap

**Pull request merged**